cypress/integration/trigger_builder_test.js

```diff
@@ -1,11 +1,263 @@
+const triggerConditionToLabel = {
+  device_connected: 'Device Connected',
+  device_disconnected: 'Device Disconnected',
+  device_error: 'Device Error',
+  device_empty_cache_received: 'Empty Cache Received',
+  incoming_data: 'Incoming Data',
+  value_change: 'Value Change',
+  value_change_applied: 'Value Change Applied',
+  path_created: 'Path Created',
+  path_removed: 'Path Removed',
+  value_stored: 'Value Stored',
+};
+
+const triggerOperatorToLabel = {
+  '*': '*',
+  '==': '==',
+  '!=': '!=',
+  '>': '>',
+  '>=': '>=',
+  '<': '<',
+  '<=': '<=',
+  contains: 'Contains',
+  not_contains: 'Not Contains',
+};
+
+const setupTriggerEditorFromSource = (trigger) => {
+  cy.get('#triggerSource')
+    .scrollIntoView()
+    .clear()
+    .type(JSON.stringify(trigger), { parseSpecialCharSequences: false });
+  cy.wait(1500);
```

Comment:
relying on `wait` to sync actions should be avoided. If typing is slower the test might fail and if it's faster you're "wasting" time waiting.
Maybe change it to something like the following
```
  const triggerSource = JSON.stringify(trigger);
  cy.get('#triggerSource')
    .scrollIntoView()
    .clear()
    .type(triggerSource, { parseSpecialCharSequences: false })
  cy.get('#triggerSource').should("have.value", triggerSource)
```

Reply:
The `cy.wait` here only waits *after* Cypress is done typing, so there should be no issue with slow or fast typing by Cypress.
However the `cy.wait(1500)` is there to wait for the app to parse the textarea input.
Doing a check as `cy.get('#triggerSource').should("have.value", triggerSource)` will not make Cypress wait long enough so that the app can parse the textarea input.